queue-exp/pifo_wfq.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# coding: utf-8 -*-
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+#
+# pifo-wfq.py
+
+"""Weighted Fair Queueing (WFQ)
+
+This scheduling algorithm is mentioned in the paper "Programmable packet
+scheduling at line rate" by Sivaraman, Anirudh, et al. It schedules flows by
+giving them a fraction of the capacity using predefined weights. In our example,
+we defined flows with an odd number to get a weight of 50 and even numbers to
+get 100.
+"""
+
+__copyright__ = """
+Copyright (c) 2021 Toke Høiland-Jørgensen <toke@toke.dk>
+Copyright (c) 2021 Frey Alfredsson <freysteinn@freysteinn.com>
+"""
+
+__license__ = """
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+from pifo_lib import Packet, Runner, Pifo
+from pifo_lib import SchedulingAlgorithm
+
+
+class Wfq(SchedulingAlgorithm):
+    """Weighted Fair Queueing (WFQ)"""
```

Comment:
Yes, I agree. It's terrible not to be able to comment on specific commits. Maybe I should also try to make more PR instead to make it more clear.

I agree it's not that much of a change to warrant the new name. I did this because the PIFO paper and examples talk about them separately, even though they look pretty much the same.

Reply:
I think it's OK to keep them separate if you think that makes things clearer; but then make the relationship between them explicit; i.e., add an `update_last_finish(flow, rank, pkt)` method to stfq and make WFQ be a subclass of STFQ that only overrides that one method...